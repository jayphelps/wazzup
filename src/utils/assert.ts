export const assert = (value: boolean, msg = 'Assertion failed') => {
  if (value === false) {
    throw new Error(msg);
  }
};
