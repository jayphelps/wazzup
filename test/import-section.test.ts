import { ast, wasm } from '../';
import { assertModule } from './utils/assertModule';

describe('Import Section', () => {
  let module;

  beforeEach(() => {
    module = new ast.Module();
  });

  test('adds function imports', () => {
    const type1 = module.addFunctionType([wasm.Type.i32], wasm.Type.i32);
    module.addImport('foo1', 'env1', 'foo', wasm.ExternalKind.Function, type1);
    assertModule(module, `
      | (module
      |   (type (;0;) (func (param i32) (result i32)))
      |   (import "env1" "foo" (func (;0;) (type 0))))
    `);

    const type2 = module.addFunctionType([wasm.Type.i64, wasm.Type.i32], wasm.Type.i32);
    module.addImport('foo2', 'env2', 'foo', wasm.ExternalKind.Function, type2);
    assertModule(module, `
      | (module
      |   (type (;0;) (func (param i32) (result i32)))
      |   (type (;1;) (func (param i64 i32) (result i32)))
      |   (import "env1" "foo" (func (;0;) (type 0)))
      |   (import "env2" "foo" (func (;1;) (type 1))))
    `);
  });

  test.skip('adds table imports', () => {
    const table1 = module.addTable([]);
    module.addImport('foo1', 'env1', 'foo', wasm.ExternalKind.Table, table1);
    assertModule(module, `
      | (module
      |   (type (;0;) (func (param i32) (result i32)))
      |   (import "env1" "foo" (table 0 anyfunc)))
    `);
  });

  test.skip('adds a single memory import', () => {
    module.addImport('foo1', 'env1', 'foo', wasm.ExternalKind.Memory, wasm.Type.none);
    assertModule(module, `
      | (module
      |   (import "env1" "foo" (memory (;0;) 1)))
    `);
  });

  // wasm MVP does not yet allow multiple memories, though the binary
  // format was designed to allow it so we prepare for it
  /*
  test('adds multiple memory imports', () => {
    module.addImport('foo1', 'env1', 'foo', wasm.ExternalKind.Memory, wasm.Type.none);
    assertModule(module, `
      | (module
      |   (import "env1" "foo" (memory (;0;) 1)))
    `);

    module.addImport('bar1', 'env2', 'bar', wasm.ExternalKind.Memory, wasm.Type.none);
    assertModule(module, `
      | (module
      |   (import "env1" "foo" (memory (;0;) 1)))
      |   (import "env2" "foo" (memory (;1;) 1))))
    `);
  });
  */

  test.skip('throws if they import multiple memories', () => {
    expect(() => {
      module.addImport('foo1', 'env1', 'foo', wasm.ExternalKind.Memory, wasm.Type.none);
      module.addImport('bar1', 'env2', 'bar', wasm.ExternalKind.Memory, wasm.Type.none);
    }).toThrowError('You can only have a single memory in wasm MVP');
  });

  test('adds global imports', () => {
    module.addImport('foo1', 'env1', 'foo', wasm.ExternalKind.Global, wasm.Type.i32);
    assertModule(module, `
      | (module
      |   (import "env1" "foo" (global (;0;) i32)))
    `);

    module.addImport('bar1', 'env2', 'bar', wasm.ExternalKind.Global, wasm.Type.i64);
    assertModule(module, `
      | (module
      |   (import "env1" "foo" (global (;0;) i32))
      |   (import "env2" "bar" (global (;1;) i64)))
    `);
  });

  test('adds global imports', () => {
    module.addImport('foo1', 'env1', 'foo', wasm.ExternalKind.Global, wasm.Type.i32);
    assertModule(module, `
      | (module
      |   (import "env1" "foo" (global (;0;) i32)))
    `);

    module.addImport('bar1', 'env2', 'bar', wasm.ExternalKind.Global, wasm.Type.i64);
    assertModule(module, `
      | (module
      |   (import "env1" "foo" (global (;0;) i32))
      |   (import "env2" "bar" (global (;1;) i64)))
    `);
  });
});
