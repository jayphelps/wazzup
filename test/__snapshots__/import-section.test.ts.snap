// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import Section adds function imports 1`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [
    Import {
      "externalKind": 0,
      "fieldName": "foo",
      "moduleName": "env1",
      "name": "foo1",
      "type": 127,
    },
  ],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [
        127,
      ],
      "returnTypes": Array [
        127,
      ],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Import Section adds function imports 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  6,
  1,
  96,
  1,
  127,
  1,
  127,
  2,
  12,
  1,
  4,
  101,
  110,
  118,
  49,
  3,
  102,
  111,
  111,
  0,
  0,
]
`;

exports[`Import Section adds function imports 3`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [
    Import {
      "externalKind": 0,
      "fieldName": "foo",
      "moduleName": "env1",
      "name": "foo1",
      "type": 127,
    },
    Import {
      "externalKind": 0,
      "fieldName": "foo",
      "moduleName": "env2",
      "name": "foo2",
      "type": 127,
    },
  ],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [
        127,
      ],
      "returnTypes": Array [
        127,
      ],
      "type": 96,
    },
    FunctionType {
      "paramTypes": Array [
        126,
        127,
      ],
      "returnTypes": Array [
        127,
      ],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Import Section adds function imports 4`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  12,
  2,
  96,
  1,
  127,
  1,
  127,
  96,
  2,
  126,
  127,
  1,
  127,
  2,
  23,
  2,
  4,
  101,
  110,
  118,
  49,
  3,
  102,
  111,
  111,
  0,
  0,
  4,
  101,
  110,
  118,
  50,
  3,
  102,
  111,
  111,
  0,
  1,
]
`;

exports[`Import Section adds global imports 1`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [
    Import {
      "externalKind": 3,
      "fieldName": "foo",
      "moduleName": "env1",
      "name": "foo1",
      "type": 127,
    },
  ],
  "memories": Array [],
  "tables": Array [],
  "types": Array [],
  "version": 1,
}
`;

exports[`Import Section adds global imports 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  2,
  13,
  1,
  4,
  101,
  110,
  118,
  49,
  3,
  102,
  111,
  111,
  3,
  127,
  0,
]
`;

exports[`Import Section adds global imports 3`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [
    Import {
      "externalKind": 3,
      "fieldName": "foo",
      "moduleName": "env1",
      "name": "foo1",
      "type": 127,
    },
    Import {
      "externalKind": 3,
      "fieldName": "bar",
      "moduleName": "env2",
      "name": "bar1",
      "type": 126,
    },
  ],
  "memories": Array [],
  "tables": Array [],
  "types": Array [],
  "version": 1,
}
`;

exports[`Import Section adds global imports 4`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  2,
  25,
  2,
  4,
  101,
  110,
  118,
  49,
  3,
  102,
  111,
  111,
  3,
  127,
  0,
  4,
  101,
  110,
  118,
  50,
  3,
  98,
  97,
  114,
  3,
  126,
  0,
]
`;

exports[`Import Section adds global imports 5`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [
    Import {
      "externalKind": 3,
      "fieldName": "foo",
      "moduleName": "env1",
      "name": "foo1",
      "type": 127,
    },
  ],
  "memories": Array [],
  "tables": Array [],
  "types": Array [],
  "version": 1,
}
`;

exports[`Import Section adds global imports 6`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  2,
  13,
  1,
  4,
  101,
  110,
  118,
  49,
  3,
  102,
  111,
  111,
  3,
  127,
  0,
]
`;

exports[`Import Section adds global imports 7`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [
    Import {
      "externalKind": 3,
      "fieldName": "foo",
      "moduleName": "env1",
      "name": "foo1",
      "type": 127,
    },
    Import {
      "externalKind": 3,
      "fieldName": "bar",
      "moduleName": "env2",
      "name": "bar1",
      "type": 126,
    },
  ],
  "memories": Array [],
  "tables": Array [],
  "types": Array [],
  "version": 1,
}
`;

exports[`Import Section adds global imports 8`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  2,
  25,
  2,
  4,
  101,
  110,
  118,
  49,
  3,
  102,
  111,
  111,
  3,
  127,
  0,
  4,
  101,
  110,
  118,
  50,
  3,
  98,
  97,
  114,
  3,
  126,
  0,
]
`;
